lib/ngx/re.lua

```diff
@@ -168,6 +152,24 @@ function _M.split(subj, regex, opts, ctx, max, res)
                 break
             end
 
+            if from == to then
+                local old_pos = ctx.pos
+
+                local from2, _, _, err2 = re_split_helper(subj, compiled,
+                                              compile_once, flags, ctx)
```

Comment:
Sure, this needs elaboration. So, considering the following string:
```
"abcd\nefgh"
```

And the following regex:
```
/^/m
```

We have a match at characters `0` and `5`. If the only thing we do is moving to the next character after the first match, we have something like the following result:
```lua
{ "a", "bcd", "e", "fgh" }
```

Because our function is not smart enough to know that an empty match does not necessarily mean we are splitting char-by-char. What we need is keep the first match (`0`), and find where the next match is (we know it'll be empty too), which will be `5`. We can then deduce our first sub-string is `1 to 4`:
```lua
{ "abcd" }
```

Otherwise, a simple increment would have cut `0 to 1`, and then `2 to 4`:
```lua
{ "a", "bcd" }
```

Simply incrementing the string pos by one would work for regexps like `/()/` were we know the next character is the next match, but not for subjects were we don't know where the next empty match will be at. At least, that's what **seems** to be the only way (insisting on the "seems") :)

On a side note, for such use cases as we are describing, we still perform 1 PCRE call per token in the subject + the last call which results in NOMATCH, such as:
```lua
ngx_re.split("abc\ndef", "^", "m") -- matches twice + NOMATCH
ngx_re.split("abc,def", ",") -- matches once + NOMATCH
```

Since we move `ctx.pos` after the look-ahead on empty matches. So there is no performance overhead, simply a slightly different logic.